pkg/synchronizer/participant.go

```diff
@@ -28,51 +28,54 @@ type participantSynchronizer struct {
 	sync.Mutex
 
 	ntpStart      time.Time
+	firstReport   time.Time
 	tracks        map[uint32]*TrackSynchronizer
 	senderReports map[uint32]*rtcp.SenderReport
 }
 
+func newParticipantSynchronizer() *participantSynchronizer {
+	return &participantSynchronizer{
+		tracks:        make(map[uint32]*TrackSynchronizer),
+		senderReports: make(map[uint32]*rtcp.SenderReport),
+	}
+}
+
 func (p *participantSynchronizer) onSenderReport(pkt *rtcp.SenderReport) {
 	p.Lock()
 	defer p.Unlock()
 
 	if p.ntpStart.IsZero() {
-		p.senderReports[pkt.SSRC] = pkt
-		if len(p.senderReports) == len(p.tracks) {
-			p.synchronizeTracks()
-		}
-		return
-	}
-
-	if t := p.tracks[pkt.SSRC]; t != nil {
+		p.initialize(pkt)
+	} else if t := p.tracks[pkt.SSRC]; t != nil {
 		t.onSenderReport(pkt, p.ntpStart)
 	}
 }
 
-func (p *participantSynchronizer) synchronizeTracks() {
-	// get estimated ntp start times for all tracks
-	estimatedStartTimes := make(map[uint32]time.Time)
-
-	// we will sync all tracks to the earliest
-	var earliestStart time.Time
-	for ssrc, pkt := range p.senderReports {
-		t := p.tracks[ssrc]
-		pts := t.getSenderReportPTS(pkt)
-		ntpStart := mediatransportutil.NtpTime(pkt.NTPTime).Time().Add(-pts)
-		if earliestStart.IsZero() || ntpStart.Before(earliestStart) {
-			earliestStart = ntpStart
+func (p *participantSynchronizer) initialize(pkt *rtcp.SenderReport) {
+	if p.firstReport.IsZero() {
+		p.firstReport = time.Now()
+	}
+
+	p.senderReports[pkt.SSRC] = pkt
+	if len(p.senderReports) < len(p.tracks) && time.Since(p.firstReport) < 5*time.Second {
+		return
+	}
+
+	// update ntp start time
+	for ssrc, report := range p.senderReports {
+		if t := p.tracks[ssrc]; t != nil {
+			pts := t.getSenderReportPTS(report)
+			ntpStart := mediatransportutil.NtpTime(report.NTPTime).Time().Add(-pts)
+			if p.ntpStart.IsZero() || ntpStart.Before(p.ntpStart) {
```

Comment:
may want to log if this adjustment is high, in SFU we drop adjustments that are more than 15 seconds. There is no good metric for how much is too high and should be ignored, but maybe logging large changes would be useful information.

Reply:
We don't make any adjustments more than 15ms. Since data is already going into gstreamer, keeping adjustments smaller than the size of an audio frame keeps things from breaking

Reply:
I used to have logs for large adjustments, but they were pretty spammy and not very useful. Now we just log the average and max drift for each track at the end

Reply:
got it, sounds good.